azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,238 @@
+                msg = "Marking the agent {0} as bad version since a downgrade was requested in the GoalState, " \
+                      "suggesting that we really don't want to execute any extensions using this version".format(CURRENT_VERSION)
+                logger.info(msg)
+                current_agent.mark_failure(is_fatal=True, reason=msg)
+            except StopIteration:
+                logger.warn(
+                    "Could not find a matching agent with current version {0} to blacklist, skipping it".format(
+                        CURRENT_VERSION))
+        else:
+            # In case of an upgrade, we don't need to exclude anything as the daemon will automatically
+            # start the next available highest version which would be the target version
+            prefix = "upgrade"
+        raise AgentUpgradeExitException("Agent update found, Exiting current process to {0} to the new Agent version {1}".format(prefix, requested_version))
+
+    @staticmethod
+    def __get_available_agents_on_disk():
+        available_agents = [agent for agent in AgentUpdateHandler.__get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    @staticmethod
+    def __get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if os.path.isdir(agent_dir)]
+
+    @staticmethod
+    def __log_event(level, msg_, success_=True):
+        if level == LogLevel.WARNING:
+            logger.warn(msg_)
+        elif level == LogLevel.ERROR:
+            logger.error(msg_)
+        elif level == LogLevel.INFO:
+            logger.info(msg_)
+        add_event(op=WALAEventOperation.AgentUpgrade, is_success=success_, message=msg_, log_event=False)
+
+    def run(self, goal_state):
+        try:
+            # Ignore new agents if update is disabled
+            if not self._autoupdate_enabled:
+                return
+
+            self._gs_id = goal_state.extensions_goal_state.id
+            agent_family = self.__get_agent_family_from_last_gs(goal_state)
+            requested_version = self.__get_requested_version(agent_family)
+
+            if requested_version == CURRENT_VERSION:
+                return
+
+            # Check if an update is allowed
+            if not self.__should_update_agent(requested_version):
+                return
+
+            msg_ = "Goal state {0} is requesting a new agent version {1}, will update the agent before processing the goal state.".format(self._gs_id, str(requested_version))
+            self.__log_event(LogLevel.INFO, msg_)
+
+            agent = self.__download_and_get_agent(goal_state, agent_family, requested_version)
+
+            if not agent.is_available:
+                msg = "Unable to download/unpack the agent version : {0} , skipping agent update".format(
```

Comment:
we need to update this message since this is not a download/unpack the agent

Reply:
All we do is download and unpack the agent when we call GuestAgent